aiida/orm/nodes/process/workflow/workflow.py

```diff
@@ -52,3 +50,23 @@ def outputs(self):
         :return: `NodeLinksManager`
         """
         return NodeLinksManager(node=self, link_type=LinkType.RETURN, incoming=False)
+
+    def validate_outgoing(self, target, link_type, link_label):
+        """Validate adding a link of the given type from ourself to a given node.
+
+        A workflow cannot 'create' Data, so if we receive an outgoing link to an unstored Data node, that means
+        the user created a Data node within our function body and tries to attach it as an output. This is strictly
+        forbidden and can cause provenance to be lost.
+
+        :param target: the node to which the link is going
+        :param link_type: the link type
+        :param link_label: the link label
+        :raise TypeError: if `target` is not a Node instance or `link_type` is not a `LinkType` enum
+        :raise ValueError: if the proposed link is invalid
+        """
+        super(WorkflowNode, self).validate_outgoing(target, link_type, link_label)
+        if link_type is LinkType.RETURN and not target.is_stored:
+            raise ValueError(
+                'trying to return an unstored Data node from a workflow is not allowed because it will cause the '
```

Comment:
this is already good. What about something like:
"Workflow '{}' tried returning an unstored Data node. This likely means new Data is being created inside the workflow. In order to preserve data provenance, use a `calcfunction` to create this data, and return its output from the workflow.".format(...)

Reply:
done